### PR TITLE
terminal: buffer the output before sending

### PR DESCRIPTION
**Pull request by paul-marechal**
In order to not overload the websocket channel opened between the
frontend and the backend we should chunk and buffer the output of
terminals.

Add `BufferingStream` that will buffer data for some time before
emitting it.

Closes https://github.com/eclipse-theia/theia/pull/11181

#### How to test

Terminals should still work.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/terminal/src/node/buffering-stream.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    protected buffer?: Buffer;
+    protected timeout?: NodeJS.Timeout;
+    protected maxChunkSize: number;
+    protected emitInterval: number;
+
+    protected onDataEmitter = new Emitter<Buffer>();
+
+    constructor(options?: BufferingStreamOptions) {
+        this.emitInterval = options?.emitInterval ?? 16; // ms
+        this.maxChunkSize = options?.maxChunkSize ?? 16384; // bytes
```
Comment (msujew) [inline packages/terminal/src/node/buffering-stream.ts:47]: I feel like 16kb is leads to too many messages being send without pauses in between. At least when trying this out (using `find ~`) on Gitpod, it still leads to the connection starvation experienced on `master`. Increasing this to 256kb does improve responsiveness massively.

**Review by msujew — COMMENTED** [review]
Looks way cleaner than the previous approach. However, I still notice resource starvation when running on non-local servers.

File: packages/terminal/src/node/buffering-stream.ts (hunk 1 above)
Comment (paul-marechal) [reply packages/terminal/src/node/buffering-stream.ts]: I wonder if could be automatically tested, I don't know how right now.

**Review by msujew — APPROVED** [review]
Great, looks good to me and cleaner than the other proposed approach 👍

**Pull request merged**